sos/report/plugins/microshift.py

```diff
@@ -0,0 +1,173 @@
+        ]
+
+        if self.get_option('add-namespaces'):
+            for nsp in self.get_option('add-namespaces').split(':'):
+                collect_regexes.append('^%s$' % nsp)
+
+        return collect_regexes
+
+    def _reduce_namespace_list(self, nsps):
+        """Reduce the namespace listing returned to just the ones we want to
+        collect from. By default, as requested by OCP support personnel, this
+        must include all 'openshift' prefixed namespaces
+
+            :param nsps list:            Namespace names from oc output
+        """
+
+        def _match_namespace(namespace):
+            """Match a particular namespace for inclusion (or not) in the
+            collection phases
+
+                :param namespace str:   The name of a namespace
+            """
+
+            for regex in self.collect_regexes:
+                if re.match(regex, namespace):
+                    return True
+            return False
+
+        self.collect_regexes = self._setup_namespace_regexes()
+
+        return list(set([n for n in nsps if _match_namespace(n)]))
+    
+    def _get_namespaces(self):
+        res = self.exec_cmd('microshift get namespaces -o custom-columns=NAME:.metadata.name --no-headers')
+        if res['status'] == 0:
+            return self._reduce_namespace_list(res['output'].split('\n'))
+        return []
+
+    def _collect_cluster_resources(self):
+        """Collect cluster-level (non-namespaced) resources
+        """
+        global_resources = [
+            'apiservices',
+            'certificatesigningrequests',
+            'clusterrolebindings',
+            'clusterroles',
+            'componentstatuses',
+            'csidrivers',
+            'csinodes',
+            'customresourcedefinitions',
+            'flowschemas',
+            'ingressclasses',
+            'logicalvolumes',
+            'mutatingwebhookconfigurations',
+            'nodes',
+            'persistentvolumes',
+            'priorityclasses',
+            'prioritylevelconfigurations',
+            'rangeallocations',
+            'runtimeclasses',
+            'securitycontextconstraints',
+            'selfsubjectaccessreviews',
+            'selfsubjectrulesreviews',
+            'storageclasses',
+            'subjectaccessreviews',
+            'tokenreviews',
+            'validatingwebhookconfigurations',
+            'volumeattachments'
+        ]
+
+        _filtered_resources = []
+
+        # Filter out any resources that might not exist.
+        for resource in global_resources:
+            res = self.exec_cmd("microshift get %s" % resource)
+            if res['status'] == 0:
+                _filtered_resources.append(resource)
+
+        _subdir = self.get_cmd_output_path(make=False)
+        self.exec_cmd(
+            'microshift inspect --kubeconfig %s --dest-dir %s %s' % (
+                self.get_option('kubeconfig'),
+                _subdir,
+                ",".join(_filtered_resources)))
```

Comment:
Calling `exec_cmd()` to generate a file in the plugin directory is fine, however there is now the paradigm in plugin execution of the `collect()` method being used for actual collections, and leaving `setup()` to _only_ define what those collections should be.

So, in this case `setup()` calling `_collect_cluster_resources()` should simply compile the list of `_filtered_resources`, and then in this plugin's `collect()` method would be something like what you have here, e.g.:

```python
    def collect(self):
        _subdir = self.get_cmd_output_path(make=False)
        _konfig = self.get_option('kubeconfig')
        self.exec_cmd(
            f"microshift inspect --kubeconfig {_konfig} --dest-dir {_subdir} "
            f"{','.join(self._filtered_resources)}"
        )

        # then the other exec_cmd() call to inspect can follow here
```

Alternatively, there is also the possibility of running this in `setup()` and getting _both_ this file and the output of the command (if for some reason this collection fails):

```python
self.add_cmd_output(
        f"microshift inspect --kubeconfig {_konfig} --dest-dir {_subdir} "
            f"{','.join(self._filtered_resources)}",
        suggest_filename='some_shorthand_name'
)

`suggest_filename` might be useful here as the standard file naming convention for `add_cmd_output()` would not be especially intuitive/helpful to users.

On that note, how does `microshift inspect` name the resulting file? I suspect we'd want this to be standardized so that automation can be used despite what `_filtered_resources` turns out to be.

Reply:
Ack.

One thing we may want to look at down the line is including these files in `sos_reports/manifest.json`. Currently, none of these files would be included in there since sos didn't directly create them. It's not a blocker for us including this plugin as the manifest is primarily for automation like Insights (which can now use tags in the manifest to locate specific files rather than relying on globs or matching files with changing command options), but just something to keep in mind.

Reply:
Finally I went with `setup` instead. `collect` is called by `_collect_manual`, which specifies in the [description](https://github.com/sosreport/sos/blob/main/sos/report/plugins/__init__.py#L3120-L3123) that it is intended for any collections outside command output and/or existing files, which is what the plugin is doing here.